### ISSUE TITLE
Istio breaks Pod's securityContext

### ISSUE THREAD

**Issue by irizzant** [istio/istio#26882]
**Bug description**
I have tried to deploy a StatefulSet with the following `securityContext`:
```yaml
imagePullSecrets: []
      initContainers: []
      securityContext:
        fsGroup: 1000
```
in an Istio enabled namespace.
When the Pod is started, I find the following on the Pod:
```yaml
securityContext:
    fsGroup: 1337
  serviceAccount: default
```

[ ] Docs
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

**Expected behavior**
I should be able to deploy a StatefulSet with the expected `fsGroup`.

**Steps to reproduce the bug**
Deploy the following test in an Istio enabled namespace:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  ports:
  - port: 80
    name: web
  clusterIP: None
  selector:
    app: nginx
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  serviceName: "nginx"
  replicas: 2
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      securityContext:
        fsGroup: 1000
      volumes:
      - name: configuration
        configMap:
          name: broker-config
      containers:
      - name: nginx
        image: k8s.gcr.io/nginx-slim:0.8
        command: ["sleep"]
        args: ["100000"]
        ports:
        - containerPort: 80
          name: web
        volumeMounts:
        - name: configuration
          mountPath: /usr/share/nginx/html/broker.xml
          subPath: broker.xml
        - name: www
          mountPath: /usr/share/nginx/html
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi
---
apiVersion: v1
data:
  broker.xml: "aaaa"
kind: ConfigMap
metadata:
  name: broker-config
```

Then exec into the container `web-0`:
kubectl exec -it web-0 -- 'ls -al /usr/share/nginx/html/'

```
drwxrwxrwx 2 root root 24 Aug 28 10:46 .
drwxr-xr-x 4 root root 30 Jul  1  2016 ..
-rw-r--r-- 1 root 1337  4 Aug 28 10:46 broker.xml
```

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
```
client version: 1.7.0
control plane version: 1.7.0
data plane version: 1.7.0 (3 proxies)
```

**How was Istio installed?**
Using Istio operator:
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
 name: istiooperator
 namespace: istio-system
spec:
  profile: demo
  components:
    ingressGateways:
    - enabled: true
      k8s:
        affinity:
          podAntiAffinity:
            preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchExpressions:
                  - key: app
                    operator: In
                    values:
                    - istio-ingressgateway
                topologyKey: kubernetes.io/hostname
              weight: 100
        service:
          ports:
          - name: tcp-activemq
            port: 61616
            targetPort: 61616
            nodePort: 32000
          - name: http2
            port: 80
            targetPort: 8080
            nodePort: 32001
          - name: https
            port: 443
            targetPort: 8443
          - name: tls
            port: 15443
            targetPort: 15443
      name: istio-ingressgateway
  values:
    sidecarInjectorWebhook:
      rewriteAppHTTPProbe: true 
    gateways:
      istio-ingressgateway:
        type: NodePort
```
**Environment where bug was observed (cloud vendor, OS, etc)**

**Comment by howardjohn** [comment]
This is due to https://github.com/kubernetes/kubernetes/issues/57923. Let me see if something has changed recently that allows us to stop doing this

**Comment by howardjohn** [comment]
We cannot.

**Comment by irizzant** [comment]
@howardjohn 
I don't see/understand exactly the relationship with https://github.com/kubernetes/kubernetes/issues/57923 I'm not using any `defaultMode` in the volumes, see manifest above.

**Comment by howardjohn** [comment]
@irizzant for Istio to work, we need to mount a secret. We also need to run as a certain user/group. In order to read that secret, we need to set securityContext on the pod due to the mentioned bug. Now that I think about it more, it *might* be possible for us to run only as UID 1337 (and not also GID)

**Comment by irizzant** [comment]
@howardjohn ok thanks for the explanation. You can specify the needed `securityContext` at container level instead of at Pod level.
This would allow you to `runAsGroup` and `runAsUser` without touching the main container/Pod permissions.

**Comment by howardjohn** [comment]
fsGroup is only available at the pod level:

```
$ k explain --recursive pod.spec.containers.securityContext
KIND:     Pod
VERSION:  v1

RESOURCE: securityContext <Object>

DESCRIPTION:
     Security options the pod should run with. More info:
     https://kubernetes.io/docs/concepts/policy/security-context/ More info:
     https://kubernetes.io/docs/tasks/configure-pod-container/security-context/

     SecurityContext holds security configuration that will be applied to a
     container. Some fields are present in both SecurityContext and
     PodSecurityContext. When both are set, the values in SecurityContext take
     precedence.

FIELDS:
   allowPrivilegeEscalation     <boolean>
   capabilities <Object>
      add       <[]string>
      drop      <[]string>
   privileged   <boolean>
   procMount    <string>
   readOnlyRootFilesystem       <boolean>
   runAsGroup   <integer>
   runAsNonRoot <boolean>
   runAsUser    <integer>
   seLinuxOptions       <Object>
      level     <string>
      role      <string>
      type      <string>
      user      <string>
   windowsOptions       <Object>
      gmsaCredentialSpec        <string>
      gmsaCredentialSpecName    <string>
      runAsUserName     <string>
```

**Comment by irizzant** [comment]
@howardjohn true, but if I understand correctly you have to mount a secret into the Istio sidecar container and for it to be usable setting `fsGroup` should not be necessary, you can use `runAsUser` and/or `runAsGroup`

**Comment by howardjohn** [comment]
@irizzant I agree with you in theory but due to https://github.com/kubernetes/kubernetes/issues/57923 it is always mounted with the wrong permissions if fsGroup is not set. I explicitly tried removing this option when I saw your issue to see if anything had changed it k8s - it still fails

**Comment by irizzant** [comment]
@howardjohn the https://github.com/kubernetes/kubernetes/issues/57923 mentions that with:
```yaml
  securityContext:
    runAsUser: 1000
    fsGroup: 1000
```
```yaml
volumes:
  - name: test-secret
    secret:
      defaultMode: 256
      secretName: test-secret
```
You get `0440` permissions.
The API documentation of `DefaultMode` states:
```
	// Mode bits to use on created files by default. Must be a value between
	// 0 and 0777.
	// Directories within the path are not affected by this setting.
	// This might be in conflict with other options that affect the file
	// mode, like fsGroup, and the result can be other mode bits set.
```
so it looks like the problem is setting `fsGroup` + `defaultMode` not `runAsUser` + `defaultMode`.
Am I missing anything?

**Comment by irizzant** [comment]
@howardjohn ok. Would it be possible to relax the `defaultMode` to allow group the read permissions and then use `runAsGroup` only?

**Comment by howardjohn** [comment]
I think ifg you put `0777` as the mode it still doesn't work, last I checked

**Comment by irizzant** [comment]
@howardjohn I think it's worth a check. Setting `defaultMode` at `0770` **without** `runAsUser` and **without** `fsGroup` but **with** `runAsGroup` should be ok

**Comment by howardjohn** [comment]
> without runAsUser

This will not work, we need to run as a special UID for the iptables traffic

**Comment by oxr463** [comment]
@howardjohn any updates on this?

**Comment by oxr463** [comment]
Ping @howardjohn

**Comment by howardjohn** [comment]
This is fixed in Kubernetes 1.19 by https://github.com/istio/istio/pull/27367. It will automatically stop setting this field if you use 1.19. I don't think its possible to fix on older versions.

**Comment by luhahn** [comment]
Hi @howardjohn ,

sorry to bother you, but I still run into this issue, even though im running on kubernetes 1.20.4 + istio 1.9.1

Istio seems to set the pods fsGroup to 1337, which blocks my container from accessing the volume.

Am I missing something here?

**Comment by scrummyin** [comment]
I may have a guess on your issue, @luhahn. I think if the pod in question has assumptions that it will run as root ([which it should not be doing](https://kubernetes.io/blog/2020/12/14/kubernetes-release-1.20-fsgroupchangepolicy-fsgrouppolicy/#allow-users-to-skip-recursive-permission-changes-on-mount)), then istio will still inject it's fsGroup of `1337`. The reason I suspect this is that the code looks to work on a merge with the existing resource. So for existing resources that **do not** have their own fsGroup set, Istio will inject one. Some applications will refuse to start if things like key or auth files have group read on them (possibly your issue).

The reason for Istio to inject it's own fsGroup (when none is present) is that fsGroup seems to have the side effect of `The permission bits are OR'd with rw-rw---- If unset, the Kubelet will not modify the ownership and permissions of any volume`([stolen from the docs here.](https://v1-20.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritycontext-v1-core)) So a fsGroup will be both injected into the groups of the user running the container **and** updated on the permission of all the mounted files to match. Which means the user groups of the process will match the filesystem groups and istio will have permissions to things like socket files with any fsGroup set. In other words istio wants either it's `1337` group or any group id (that could already be present as part of the underlying pod).

A solution here @luhahn, might be to take a look at the pod which has the issue and ensure that it works when not running as root. I do wonder if there could be some flagging or detection by Istio of containers running as root to not inject an fsGroup (as they would already have root perms, and would not need additional ones), but this feels like a bit of a hack to work around pods working against the k8s security mentality of non-root containers.

Side note, I'm currently experiencing an issue with [mongodb's keyfile](https://docs.mongodb.com/manual/tutorial/deploy-replica-set-with-keyfile-access-control/#create-a-keyfile) that requires no group perms be set on it. However, other datastores like [postgresql allow for cases where group perms might be used](https://www.postgresql.org/docs/current/ssl-tcp.html), which is more friendly things like the fsGroup or [often backup software](https://paquier.xyz/postgresql-2/postgres-11-group-access/). So this might be an issue of who should actually be dealing with a certain pod not allowing fsGroup. Is that the underlying software project (mongodb in my case), the container/operator maintainers (mongo-operator), kubernetes (via features like fsGroupChangePolicy), Istio, or maybe a little bit of all those teams.

**Comment by resamaraschi** [comment]
@luhahn you need to set "ENABLE_LEGACY_FSGROUP_INJECTION" to false for your istio pilot-agent in order to disabling fsGroup injection.

**Comment by Keramblock** [comment]
Could you please provide how could I use that variable, when I deploing that using operator helm chart?

**Comment by resamaraschi** [comment]
you can set the env in your istioOperator CustomResource:
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: ...
spec:
  tag: "1.9.4-distroless"
...
  components:
    pilot:
      k8s:
        env:
        - name: ENABLE_LEGACY_FSGROUP_INJECTION
          value: "false"

```

**Comment by Keramblock** [comment]
Thanks a lot! =)

**Comment by ahawtho** [comment]
We have run into this issue also. From what I understand, the summary of the problem is that the Istio secret is not mounted with the appropriate permissions in k8s < 1.19 due to https://github.com/kubernetes/kubernetes/issues/57923 . Istio triggers this because it sets `runAsUser` and `runAsGroup`, and as a result, there's a permission issue with the secret. But after reading through the comments on https://github.com/kubernetes/kubernetes/issues/57923 , it seems as though the problem reported there is that the secret is mounted with *more* permissions than are strictly necessary. The complaint in that issue is that the group read bit `0010` is set when the mode `0400` is requested by the container, so the effective permissions are `0440`.

My understanding is that the istio sidecar container should inherit the supplemental group provided by any `fsGroup` setting when it sets `runAsUser`. Perhaps the ship has already sailed, but it seems like if `fsGroup` is set initially, Istio should be able to avoid overwriting `fsGroup` and leave `runAsGroup` unset and still be able to read the secret. If the code using the secret requires certain permissions, it might be necessary for it to copy the file to a new file with appropriate permissions, but it should be possible to solve this way.

The impact for us is high because we run on EKS and leverage the IAM roles assigned to `ServiceAccount` objects and populated by another admission controller provided by EKS. The EKS admission controller mounts the web identity tokens generated for a given IAM role with mode `0600`. K8s updates this to `0440` for a read-only mount and when providing an `fsGroup`, but when Istio changes the `fsGroup`, our services are unable to read the EKS web identity token since they rely on a specific group that we set in the `fsGroup` setting.

We didn't see this issue in Istio 1.5, but now that we're trying to catch up to current, and going through the difficult 1.5 -> 1.6 transition, this impacts virtually all our services, which depend on the token having group ownership other than `1337`. We're on EKS 1.16 which does not have the fix from 1.19 . This issue leaves us between a bit of a rock and a hard place.

Any thoughts, corrections, or help here?

**Comment by yann-soubeyrand** [comment]
Hi @howardjohn, if I understand this PR https://github.com/istio/istio/pull/27367 correctly, on a Kubernetes version 1.21 cluster running Istio version 1.11.4 (installed via the operator) I should get the `ENABLE_LEGACY_FSGROUP_INJECTION` environment variable set to false on my istiod deployment, right? But this is not the case in my cluster. Did I missed something? Could I break something if I manually set this environment variable (by setting `spec.components.pilot.k8s.env` in my IstioOperator CR)?
